B plus tree

arv opened the pull request:
This implements a B+ Tree

Fixes rocicorp/replicache#596 .

Left to do:
 - [x] diff
 - [ ] Change default size
 - [x] Make on chunk layout a bit more compact (no need to store large type property)
 - [ ] "cache" size so that the size is not recomputed all the time

arv committed:
WIP: Put works OK with fixed size of 1

arv committed:
Del witout rebalancing

arv committed:
Del with partiotioning

arv committed:
Add tests for temp hash

arv committed:
partionn top level

arv committed:
Partition is splitting correctly now

arv committed:
Cleanup test

arv committed:
broken

arv committed:
del returns new node

arv committed:
more

arv committed:
more

arv committed:
Time to start doing real size

arv committed:
more

arv committed:
Cleanup

and use getSize

arv committed:
coment about structured clone source code

arv committed:
scan

arv committed:
Better size calc

arv committed:
More size tests

arv committed:
no nanoid

arv committed:
more

arv committed:
Use Hash type consistently

Use smaller chunk data

aboodman commented:
Code size increase?

Also:
* impact on existing perf tests?
* we should expand the existing subscribe tests to cover our write/sub/read target here: rocicorp/mono#134. This bug should demonstrate us meeting that target using `memStore`.

aboodman commented:
(in-depth review coming)

arv committed:
WIP: Work on size of chunk

aboodman reviewed (approved):
I am not the right person to review the logic in detail. I would really like @phritz to take a pass when he has time, but low priority (can be after you move forward).

Structurally this LGTM, but I have a few questions about caching node impl instances and the layout of the pod structures.

arv committed:
Code review feedback and include chunk header size in computation

arv committed:
cleanup

arv committed:
Add diff

This uses scan to implement diff. Not the most efficient way to do it
but it unblocks us

arv committed:
More cleanup

arv committed:
Integrated

arv committed:
add missing await

arv committed:
Add diff

This is based on Map diff from Noms

Main difference is that we do not need to use edit distance since our
entries are already sorted which allows a O(n+m) solution instead of
O(n*m)

arv committed:
Remove type

remove varint consideration for size of value

arv committed:
Remove debug comment

arv committed:
Cleanup computeSplices

arv committed:
Use tuple for chunk data

arv committed:
Merge branch 'main' into b-plus-tree
